crates/sui-core/src/checkpoints/mod.rs

```diff
@@ -895,28 +895,25 @@ impl CheckpointSignatureAggregator {
             );
             return Err(());
         }
-        match self.signatures.insert(author, signature) {
+        match self.signatures.insert(signature) {
             InsertResult::RepeatingEntry { previous, new } => {
                 if previous != new {
                     warn!("Validator {:?} submitted two different signatures for checkpoint {}: {:?}, {:?}", author.concise(), self.summary.sequence_number, previous, new);
                 }
                 Err(())
             }
-            InsertResult::QuorumReached(values) => {
-                let signatures = values.values().cloned().collect();
-                match AuthorityWeakQuorumSignInfo::new_from_auth_sign_infos(
-                    signatures,
-                    self.signatures.committee(),
-                ) {
-                    Ok(aggregated) => Ok(aggregated),
-                    Err(err) => {
-                        error!(
-                            "Unexpected error when aggregating signatures for checkpoint {}: {:?}",
-                            self.summary.sequence_number, err
-                        );
-                        Err(())
-                    }
-                }
+            InsertResult::Failed { error } => {
+                warn!(
+                    "Failed to aggregate new signature from validator {:?} for checkpoint {}: {:?}",
+                    author.concise(),
+                    self.summary.sequence_number,
+                    error
+                );
+                Err(())
+            }
+            InsertResult::QuorumReachedWithCert(cert) => Ok(cert),
+            InsertResult::QuorumReached => {
+                unreachable!("insert should only return QuorumReachedWithCert");
```

Comment:
I think you can avoid unreachable here by adding generic parameter to InsertResult<R>.
`insert_generic` would return InsertResult<()> and `insert` would return `InsertResult<Cert>`. This way only one enum element is needed(QuorumReached)

Reply:
Updated. It does work though the translation from InsertResult<()> to InsertResult<Cert> is a bit ugly.